caper/resource_analysis.py

```diff
@@ -0,0 +1,338 @@
+        i.e. max_mem and max_disk.
+        Therefore, y_vec_1 should look like:
+            [
+                STATS_MAX_MEM1,
+                STATS_MAX_MEM2,
+                ...
+            ]
+        y_vec_2 should look like:
+            [
+                STATS_MAX_DISK1,
+                STATS_MAX_DISK2,
+                ...
+            ]
+        They are two separate problems.
+
+        Args:
+            in_file_vars:
+                List of input file vars. [IN_FILE_VAR1, ...].
+                Matching input file will be included in x_matrix of analysis.
+                See `input_file_sizes` in CromwellMetadata.gcp_monitor.__doc__ for details.
+                If None or False, then all input files will be used to make x_matrix.
+            reduce_in_file_vars:
+                Python function to reduce x_vector. e.g. sum, max.
+                If None, then no reduction of vector x.
+                If you don't reduce x, make sure that you have enough number of data (>=dim(x)).
+                Some solver can fail due to lack of data. i.e. dim(problem, solver) > dim(data).
+                For example, even a linear solver will fail with only one dataset.
+            target_resources:
+                Keys (in dot notation) to make vector y. e.g. ('stats.max.mem', 'stats.max.disk').
+                See CromwellMetadata.gcp_monitor.__doc__ to find available keys.
+                One vector y for each key, which means one solving for each key.
+            plot_pp:
+                Matplotlib's PDF backend PdfPages to write plots on multiple pages
+                (task per page).
+
+        Returns:
+            coeffs:
+                Analysis result.
+                e.g. (coeffs, intercept) for linear regression.
+            Plot object:
+                Matplotlib plot object.
+        """
+        import numpy as np
+
+        result = {}
+
+        in_file_vars_found = set()
+        x_data = defaultdict(list)
+        y_data = defaultdict(list)
+
+        logger.info('Analyzing task={task}'.format(task=task_name))
+        # first look at task's optional/empty input file vars across all workflows
+        # e.g. SE (single-ended) pipeline runs does not have fastqs_R2
+        # but we want to mix both SE/PE (paired-ended) data.
+        # so need to look at all workflows to check if optional/empty var is
+        # actully a file var.
+        for task in self._tasks:
+            if not fnmatch.fnmatchcase(task['task_name'], task_name):
+                continue
+
+            for in_file_var, in_file_size in task['input_file_sizes'].items():
+                if in_file_vars and in_file_var not in in_file_vars:
+                    continue
+
+                if in_file_size:
+                    in_file_vars_found.add(in_file_var)
+
+        for task in self._tasks:
+            if not fnmatch.fnmatchcase(task['task_name'], task_name):
+                continue
```

Comment:
I question the need to iterate over self._tasks twice. Also the `fnmatch` is repetitive with line 201 above.

Reply:
Yeah, we need to look at all tasks (across all workflows) first to find all non-empty input file variables.
e.g. input variable `fastqs_R2` is empty for task `align` of single-ended workflows. So Caper does cannot know that it is a `File` type. Caper (autouri) only recognizes abspath, URLs or URIs (starting with `gs://`) as `File`s. It needs to look at other workflows with non-empty `fastqs_R2`.

That's a good point. Added `matched_task_resources` to store matched tasks to avoid repetitive calling of `fnmatch`.

Reply:
Thanks for the explanation, makes sense.